### PR TITLE
Use CircleCI to run the unit tests

### PR DESCRIPTION
**Pull request by JordanP**

### CODE REVIEW

**Comment by NicolasT** [comment]
Commit message of 362324f should be fixed: `setup.py: Fail gracefully when Git tags can't be found`.

**Comment by JordanP** [comment]
Yup. Will do. I am working on generating a proper Xunit report now.

**Comment by JordanP** [comment]
Disappointing: all I get from the artifacts is "Your build ran 389 tests in unknown with 0 failures."

**Comment by JordanP** [comment]
https://circleci.com/gh/scality/ScalitySproxydSwift/15 shows the output of CircleCI when a test failed

**Comment by NicolasT** [comment]
Looks OK, isn't it?

**Comment by JordanP** [comment]
Yes.

**Comment by NicolasT** [comment]
I'm a bit worried by the '999' hack if no Git tags can be found. What if someone installs the package from the GitHub-provided zip-file from the master branch on a non-CI system? Wouldn't this system be stuck with 'that 999 version'? As in: until now this was an (indirect) protection against installing master from zips.

**Comment by JordanP** [comment]
You're right. In a venv I tested "pip install https://github.com/scality/ScalitySproxydSwift/archive/circleci.zip" and got "Successfully installed swift-scality-backend-999"

**Comment by JordanP** [comment]
In the setup.py we could detect if we run in CircleCI based on the "CIRCLECI" env variable  ( https://circleci.com/docs/environment-variables )

**Comment by NicolasT** [comment]
Well, I was partially wrong: that '999' doesn't come from the failing Git lookup, but is the default when the source is not a Git repository... https://github.com/scality/ScalitySproxydSwift/blob/87222c0e85c0d4bd8759697c38e42fd35705afde/setup.py#L58

Maybe a better solution would then be to use 'yearmonthdayhourminutesecond' (20151119111634) as the version instead of '999'.
Or bail out when installation is attempted from a non-release Zip?
Or run 'git fetch --tags' during the CircleCI build?

**Comment by JordanP** [comment]
Let's talk about it in person next time.

**Comment by NicolasT** [comment]
One more concern: won't the xunit filename change impact our Jenkins job? Will those still find the test results where they expect them?

**Comment by JordanP** [comment]
I'd like to remove the corresponding jenkins job. It's inefficient (it spawns ~20 VMs) and slow. 

We now have a coverage report. For example: https://circle-artifacts.com/gh/scality/ScalitySproxydSwift/23/artifacts/0/tmp/circle-junit.6AOB5kg/coverage/index.html

**Comment by NicolasT** [comment]
Jenkins runs on multiple OS'es, which seems like a big plus to me. Remember the issues we had with `urllib3`.

**Comment by JordanP** [comment]
Yeah but urllib3 is now a dependency of scality-sproxyd-client. And even so, it should be mocked everywhere in unit tests. We still have functional tests managed by Jenkins.

File: circle.yml [hunk 1]
```diff
@@ -0,0 +1,19 @@
+checkout:
+  post:
```
Comment (NicolasT) [inline circle.yml]: It might be worth leaving a little comment here explaining why this is necessary.

File: circle.yml [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+dependencies:
+  override:
+    - pip install tox coverage
+    - pyenv local 2.7.10 2.6.8
+
+test:
+  pre:
+    - coverage erase
+  override:
+    - tox -v --recreate -- --with-xunit
```
Comment (NicolasT) [inline circle.yml]: Do we always want to recreate? CircleCI has a fairly useful caching feature, which could speed up builds quite a bit.

Comment (JordanP) [inline circle.yml]: Ok. I hope CircleCI is smart enough to invalidate the cache if requirements are changed.

File: circle.yml [hunk 3]
```diff
@@ -0,0 +1,21 @@
+checkout:
+  post:
+    # Fetch the complet repository. Required to get the tags and
```
Comment (NicolasT) [inline circle.yml]: `s/complet/complete`

**Comment by NicolasT** [comment]
I noticed you removed the `remove` option from the Tox run. For now this won't change anything, since all jobs run in a fresh container. You'll want to add `.tox` to the CircleCI cache: see https://circleci.com/docs/configuration#cache-directories

**Comment by JordanP** [comment]
Without cache: 3min 07; 
With cache: 2min 04.

It seems to work.

**Comment by NicolasT** [comment]
:+1: Good job, LGTM.

**Pull request merged**